projects/charger/src/generic_can_uart.c

```diff
@@ -0,0 +1,106 @@
+#include "generic_can_uart.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "can_uart.h"
+#include "generic_can.h"
+#include "status.h"
+
+static CanUart s_can_uart;
+static GenericCanInterface s_interface;
+
+// CanUartRxCb
+static void prv_generic_can_uart_rx_handler(const CanUart *can_uart, uint32_t id, bool extended,
+                                            const uint64_t *data, size_t dlc, void *context) {
+  (void)can_uart;
+  GenericCanUart *gcu = context;
+  for (size_t i = 0; i < NUM_GENERIC_CAN_UART_RX_HANDLERS; i++) {
+    if (id == gcu->rx_storage[i].id && gcu->rx_storage[i].rx_handler != NULL &&
+        gcu->rx_storage[i].enabled) {
+      const GenericCanMsg msg = {
+        .id = id,
+        .extended = extended,
+        .data = *data,
+        .dlc = dlc,
+      };
+      gcu->rx_storage[i].rx_handler(&msg, gcu->rx_storage[i].context);
+      break;
+    }
+  }
+}
+
+// tx
+static StatusCode prv_tx(const GenericCan *can, const GenericCanMsg *msg) {
+  GenericCanUart *gcu = (GenericCanUart *)can;
+  if (gcu->base.interface != &s_interface || gcu->can_uart != &s_can_uart) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "GenericCan not aligned to GenericCanUart.");
+  }
+
+  return can_uart_req_slave_tx(gcu->can_uart, msg->id, msg->extended, &msg->data, msg->dlc);
+}
+
+// register_rx
+static StatusCode prv_register_rx(GenericCan *can, GenericCanRx rx_handler, uint32_t id,
+                                  void *context) {
+  GenericCanUart *gcu = (GenericCanUart *)can;
```

Comment:
I feel like there's a lot of shared code for the handlers between your implementations.

Reply:
Done. I split the RX handling into a separate module. The TX is less generic.